test/e2e/virtual-routes.test.js

```diff
@@ -0,0 +1,293 @@
+          routes,
+        },
+      });
+
+      await navigateToRoute(page, '/multiple/matches');
+
+      const titleElm = page.locator('#main h1');
+      await expect(titleElm).toContainText('First Match');
+    });
+
+    test('prefer virtual route over a real file, if a virtual route exists', async ({
+      page,
+    }) => {
+      const routes = {
+        '/': '# Virtual Homepage',
+      };
+
+      await docsifyInit({
+        markdown: {
+          homepage: '# Real File Homepage',
+        },
+        config: {
+          routes,
+        },
+      });
+
+      const titleElm = page.locator('#main h1');
+      await expect(titleElm).toContainText('Virtual Homepage');
+    });
+
+    test('fallback to default routing if no route was matched', async ({
+      page,
+    }) => {
+      const routes = {
+        '/a': '# A',
+        '/b': '# B',
+        '/c': '# C',
+      };
+
+      await docsifyInit({
+        markdown: {
+          homepage: '# Real File Homepage',
+        },
+        config: {
+          routes,
+        },
+      });
+
+      await navigateToRoute(page, '/d');
+
+      const mainElm = page.locator('#main');
+      await expect(mainElm).toContainText('404 - Not found');
+    });
+
+    test('skip routes that returned a falsy value that is not a boolean', async ({
+      page,
+    }) => {
+      const routes = {
+        '/multiple/(.+)': () => null,
+        '/multiple/(.*)': () => undefined,
+        '/multiple/.+': () => 0,
+        '/multiple/.*': () => '# Last Match',
+      };
+
+      await docsifyInit({
+        config: {
+          routes,
+        },
+      });
+
+      await navigateToRoute(page, '/multiple/matches');
+
+      const titleElm = page.locator('#main h1');
+      await expect(titleElm).toContainText('Last Match');
+    });
+
+    test('abort virtual routes and not try the next one, if any matched route returned an explicit "false" boolean', async ({
+      page,
+    }) => {
+      const routes = {
+        '/multiple/(.+)': () => false,
+        '/multiple/(.*)': () => "# You Shouldn't See Me",
+      };
+
+      await docsifyInit({
+        config: {
+          routes,
+        },
+      });
+
+      await navigateToRoute(page, '/multiple/matches');
+
+      const mainElm = page.locator('#main');
+      await expect(mainElm).toContainText('404 - Not found');
+    });
+
+    test('skip routes that are not a valid string or function', async ({
+      page,
+    }) => {
+      const routes = {
+        '/multiple/(.+)': 123,
+        '/multiple/(.*)': false,
+        '/multiple/.+': null,
+        '/multiple/..+': [],
+        '/multiple/..*': {},
+        '/multiple/.*': '# Last Match',
+      };
+
+      await docsifyInit({
+        config: {
+          routes,
+        },
+      });
+
+      await navigateToRoute(page, '/multiple/matches');
+
+      const titleElm = page.locator('#main h1');
+      await expect(titleElm).toContainText('Last Match');
+    });
+  });
+});
+
+/**
```

Comment:
Last nitpick (promise)...

Can we move this after the imports and before the first `test()` block? This has been the general pattern for local helpers in other test files.

Reply:
np, done & done